spinalcordtoolbox/reports/qc.py

```diff
@@ -419,12 +424,17 @@ def update_description_file(self, dimension):
         }
         logger.debug('Description file: %s', self.qc_params.qc_results)
         results = []
-        if os.path.isfile(self.qc_params.qc_results):
-            results = json.load(open(self.qc_params.qc_results, 'r'))
-        results.append(output)
-        json.dump(results, open(self.qc_params.qc_results, "w"), indent=2)
+        with portalocker.Lock(self.qc_params.qc_results) as lck_qc_file:
+            with touchopen(lck_qc_file.name, 'r+') as edit_lck_qc:
```

Comment:
It may work but I'm not a huge fan of this manipulation as it's opening the same file twice and makes the code complicated.
Can you encapsulate this in a prettier-looking class?

Portalocker provides portable low-level functions so we could use those.

I believe you could do something like:
```python
class open_with_exclusive_lock(object):
    """
    """
    def __init__(self, filename, mode):
        ...
    def __enter__(self):
        fd = os.open(...)
        self._f = os.fdopen(...)
        portlalocker.lock(self._f)
        return self._f
  
    def __exit__(self, exc_type, exc_val, exc_tb):
        portalocker.unlock(self._f)
```

Reply:
(and this can go in sct_utils.py)